igvm/hypervisor_preferences.py

```diff
@@ -1,63 +1,93 @@
+
+Copyright (c) 2020 InnoGames GmbH
 """
-# This module contains the preferences as simple classes.  We try to keep
-# them reusable, even though most of them are not reused.  Some of the classes
-# are so simple that they could as well just be a function, but kept
-# as classes to have a consistent style.
-
 from logging import getLogger
-
-from igvm.utils import LazyCompare
+from typing import Union, List
 
 log = getLogger(__name__)
 
 
-class InsufficientResource(object):
-    """Check a resource of hypervisor would be sufficient"""
-    def __init__(self, hv_attribute, vm_attribute, multiplier=1, reserved=0):
-        self.hv_attribute = hv_attribute
-        self.vm_attribute = vm_attribute
-        self.multiplier = multiplier  # TODO Use identical units in Serveradmin
-        self.reserved = reserved
-
-    def __repr__(self):
+class HypervisorPreference:
+    """The base class for all HV preferences."""
+
+    def get_score(self, vm, hv) -> Union[float, bool]:
+        """Calculates a preference value to indicate how good a HV fits.
+
+        @param vm: The VM object to check against the HV.
+        @param hv: The HV object to check against the VM.
+        @return: A float or a bool. The bool is just for convenience and will
+                 be converted to a float. A value of 0.0 means the HV does not
+                 fulfill the requirement at all. But a value of 1.0 means it is
+                 a very good candidate for this particular preference.
+        @rtype: Union[float, bool]
+        """
+        raise NotImplementedError('get_preference is not implemented')
+
+
+class InsufficientResource(HypervisorPreference):
+    """Check whether a resource of a hypervisor would be sufficient."""
+
+    def __init__(
+        self,
+        hv_attribute: str,
+        vm_attribute: str,
+        multiplier: int = 1,
+        reserved: int = 0,
+    ) -> None:
+        self.hv_attribute: str = hv_attribute
+        self.vm_attribute: str = vm_attribute
+        # TODO Use identical units in Serveradmin
+        self.multiplier: int = multiplier
+        self.reserved: int = reserved
+
+    def __repr__(self) -> str:
         args = repr(self.hv_attribute)
         if self.reserved:
             args += ', reserved=' + repr(self.reserved)
 
         return '{}({})'.format(type(self).__name__, args)
 
-    def __call__(self, vm, hv):
+    def get_score(self, vm, hv) -> Union[float, bool]:
         # Treat freshly created HVs always passing this check
         if not hv.dataset_obj[self.hv_attribute]:
-            return False
+            return True
 
-        total_size = hv.dataset_obj[self.hv_attribute]
+        # Calculate the remaining "size" of the resource
+        total_size = hv.dataset_obj[self.hv_attribute] * self.multiplier
         vms_size = sum(
-            vm[self.vm_attribute] * self.multiplier
+            vm[self.vm_attribute]
             for vm in hv.dataset_obj['vms']
+            if vm['state'] != 'retired'
```

Comment:
Does that work for all kind of resources ? If we assume retired VMs are shutdown (not guaranteed) and we are talking about CPUs and Memory the resources are not used but can we really always be certain a retired VM is shutdown and a resource is not used (I lack a good example but imagine we would calculate disk space this way).

Reply:
i get your point. first in our cases we do expect retired objects to not be in use anymore. you are right that it is not guaranteed but i don't see this as a real problem here because we do make later "hard" checks against libvirt before using a hypervisor. that way we are safe against all inconsistencies. these preferences just generate ordered hv "suggestions". whether a vm really fits on it is not the purpose here. is this fine for you?

Reply:
Yes I think you are right this is safe enough and if it really fits is a second layer.